### DISCUSSION TITLE
Sign with GPG

### DISCUSSION THREAD

**Discussion by JCHacking** [jimporter/mike#228] [Ideas]
I have several repositories with the rule to only allow signed commits.

I see that with mike even if I configure the GPG, the commit that it generates is not signed, so it is rejected by GitHub.
```
error: failed to push branch doc to origin:
  warning: not sending a push certificate since the receiving end does not support --signed push
  ***
  remote: - Commits must have verified signatures.
  remote:   Found 1 violation:
  ***
```

Is there any possibility of implementing it? Any advice on how to approach it if I end up implementing it myself in this tool?

**Comment by jimporter** [comment]
mike just uses `git fast-import` to generate commits, so it does whatever that Git subcommand does. Maybe that supports GPG-signing, and maybe not (the Git documentation isn't really clear about this). If Git doesn't support that, then you can always just amend the commit generated by mike to add a signature. Something like `git checkout gh-pages && git commit --amend -S` would probably do the trick.

**Reply by JCHacking** [reply]
Good idea! I hadn't thought of that

Maybe it could be implemented automatically in `mike` with the postcommit concept, to read if the `commit.gpgsign` signature property exists and if so do the `amend`?

EDIT: 
It seems that work has started here, but I see that it is not merged or finished [Git Sign Commit fast-import and fast-export](https://patchwork.kernel.org/project/git/patch/20210430232537.1131641-6-lukeshu@lukeshu.com/#24152321) 
I understand then that it does not implement the `fast-import` commit signature, otherwise it would not have wanted to implement this.

**Comment by JCHacking** [comment]
In case anyone is interested, or is in the same need, the solution that I have applied was:

Previous code:
```bash
mike deploy --branch doc --push "${VERSION}"
mike set-default --branch doc --push "${VERSION}"
```

New code with the solution applied:
```bash
mike deploy --branch doc "${VERSION}"
(git checkout doc && git commit --amend --no-edit) || { git checkout -; exit 1; }; git checkout -
mike set-default --branch doc "${VERSION}"
(git checkout doc && git commit --amend --no-edit) || { git checkout -; exit 1; }; git checkout -
git push --set-upstream origin doc
```
This will make that after each change, an amend is made, for that I have to follow the steps:
- checkout doc branch
- amend
- checkout original branch/tag
- finally manually upload both changes, using upstream to create the branch in case it does not exist.
In case of any failure, return to the original branch and exit with code 1.

**Comment by bernardomaciel-apheris** [comment]
There seems to be support for signing with `fast-import` now: https://git-scm.com/docs/git-fast-import#_gpgsig .

Any chance we can have this in `mike`?

**Reply by bernardomaciel-apheris** [reply]
@jimporter apologies for the direct ping, but WDYT?

Happy to open a new discussion if needed

**Reply by jimporter** [reply]
Given the following comment in the Git docs, I think it's probably a bit premature to try and use this in mike:

> Note: This is highly experimental and the format of the gpgsig command may change in the future without compatibility guarantees.

**Reply by bernardomaciel-apheris** [reply]
Got it - if I wanted to give it a stab, where do you suggest I start?

**Reply by jimporter** [reply]
Does amending the commit as described above not work? That's probably the most reliable way to do it for now, since otherwise you'd have to figure out what key to add, get its data, and then include it via `fast-import`. Amending the commit is a shell one-liner, on the other hand.

**Reply by bernardomaciel-apheris** [reply]
It works, but it would be much nicer to do `mike deploy --sign --push ...` ...
